### ISSUE TITLE
Q: Combine multiple losses

### ISSUE THREAD

**Issue by s1ddok** [tensorflow/swift-apis#724]
Hi! I'm on my way of implementing CycleGAN in Swift for TensorFlow, but ran into an issue I can't resolve myself. I know GitHub issues is not the place to ask questions, I would love to be pointed with an appropriate place.

I wrote the code to calculate multiple losses:

```swift
            let (fakeY, fakeYBackprop) = generatorG.appliedForBackpropagation(to: realX)
            let (cycledX, cycledXBackprop) = generatorF.appliedForBackpropagation(to: fakeY)
            
            let (fakeX, fakeXBackprop) = generatorF.appliedForBackpropagation(to: realY)
            let (cycledY, cycledYBackprop) = generatorG.appliedForBackpropagation(to: fakeX)
            
            // sameX and sameY are used for identity loss.
            let (sameX, sameXBackprop) = generatorF.appliedForBackpropagation(to: realX)
            let (sameY, sameYBackprop) = generatorG.appliedForBackpropagation(to: realY)
            
            let (discRealX, discRealXBackprop) = discriminatorX.appliedForBackpropagation(to: realX)
            let (discRealY, discRealYBackprop) = discriminatorY.appliedForBackpropagation(to: realY)
            
            let (discFakeX, discFakeXBackprop) = discriminatorX.appliedForBackpropagation(to: fakeX)
            let (discFakeY, discFakeYBackprop) = discriminatorY.appliedForBackpropagation(to: fakeY)
            
            let (genGLoss, genGGradient) = TensorFlow.valueWithGradient(at: discFakeY) {
                sigmoidCrossEntropy(logits: $0, labels: Tensorf.one.broadcasted(like: $0))
            }
            let (genFLoss, genFGradient) = TensorFlow.valueWithGradient(at: discFakeX) {
                sigmoidCrossEntropy(logits: $0, labels: Tensorf.one.broadcasted(like: $0))
            }

            let (cycleLossX, cycleLossXGradient) = TensorFlow.valueWithGradient(at: realX, cycledX) {
                abs($0 - $1).mean() * lambdaL1
            }
            let (cycleLossY, cycleLossYGradient) = TensorFlow.valueWithGradient(at: realY, cycledY) {
                abs($0 - $1).mean() * lambdaL1
            }
```

and now I'm struggling to combine them into one. In Python tutorial they just do it like this:

```
    gen_g_loss = generator_loss(disc_fake_y)
    gen_f_loss = generator_loss(disc_fake_x)
    
    total_cycle_loss = calc_cycle_loss(real_x, cycled_x) + calc_cycle_loss(real_y, cycled_y)
    
    # Total generator loss = adversarial loss + cycle loss
    total_gen_g_loss = gen_g_loss + total_cycle_loss + identity_loss(real_y, same_y)
    total_gen_f_loss = gen_f_loss + total_cycle_loss + identity_loss(real_x, same_x)

    disc_x_loss = discriminator_loss(disc_real_x, disc_fake_x)
    disc_y_loss = discriminator_loss(disc_real_y, disc_fake_y)
```

but I'm not sure how do I combine `multiple gradients, losses and backpropagators into single model gradient to be passed into Adam optimizer`. I would greatly appreciate any advice!

**Comment by saeta** [comment]
> I know GitHub issues is not the place to ask questions, I would love to be pointed with an appropriate place.

Thanks for flagging that we don't have this documented anywhere! 
 I just put together: https://github.com/tensorflow/swift-apis/wiki/Getting-Help. Is that helpful? // Any suggested edits?

**Comment by s1ddok** [comment]
@saeta great, thanks! I will try to redirect the question there.

Btw did you consider creating a dedicated subforum on forums.swift.org? I feel like mailing lists are a bit scary especially for newcomers like me.

**Comment by saeta** [comment]
Hehe, great question. There are plans to get a discourse instance setup for TensorFlow, and we were planning a subforum there. But it hasn't finalized yet, so we just have the mailing list for now. :-/ Thanks for the feedback, and for the question on the mailing list!
